src/feature_extraction.py

```diff
@@ -477,6 +477,112 @@ def persist(self, filepath):
         joblib.dump(self.features, filepath)
 
 
+class PreviousApplicationFeatures(BaseTransformer):
+    def __init__(self, **kwargs):
+        self.features = None
+
+    @property
+    def feature_names(self):
+        feature_names = list(self.features.columns)
+        feature_names.remove('SK_ID_CURR')
+        return feature_names
+
+    def fit(self, X, prev_applications, **kwargs):
+        features = pd.DataFrame({'SK_ID_CURR': prev_applications['SK_ID_CURR'].unique()})
+
+        prev_applications_sorted = prev_applications.sort_values(['SK_ID_CURR', 'DAYS_DECISION'])
+        prev_applications_last_3 = prev_applications_sorted.groupby(by=['SK_ID_CURR']).tail(3)
+        prev_applications_last_5 = prev_applications_sorted.groupby(by=['SK_ID_CURR']).tail(5)
+
+        group_object = prev_applications_sorted.groupby(by=['SK_ID_CURR'])['SK_ID_PREV'].count().reset_index()
+        group_object.rename(index=str,
+                            columns={'SK_ID_PREV': 'previous_application_number_of_prev_application'},
+                            inplace=True)
+        features = features.merge(group_object, on=['SK_ID_CURR'], how='left')
+
+        group_object = prev_applications_sorted.groupby(by=['SK_ID_CURR'])['CNT_PAYMENT'].last().reset_index()
+        group_object.rename(index=str,
+                            columns={'CNT_PAYMENT': 'previous_application_term_of_previous_credit'},
+                            inplace=True)
+        features = features.merge(group_object, on=['SK_ID_CURR'], how='left')
+
+        group_object = prev_applications_last_3.groupby(by=['SK_ID_CURR'])['CNT_PAYMENT'].mean().reset_index()
+        group_object.rename(index=str,
+                            columns={'CNT_PAYMENT': 'previous_application_term_of_last_3_credit_mean'},
+                            inplace=True)
+        features = features.merge(group_object, on=['SK_ID_CURR'], how='left')
+
+        group_object = prev_applications_last_5.groupby(by=['SK_ID_CURR'])['CNT_PAYMENT'].mean().reset_index()
+        group_object.rename(index=str,
+                            columns={'CNT_PAYMENT': 'previous_application_term_of_last_5_credit_mean'},
+                            inplace=True)
+        features = features.merge(group_object, on=['SK_ID_CURR'], how='left')
+
+        group_object = prev_applications_sorted.groupby(by=['SK_ID_CURR'])['DAYS_DECISION'].last().reset_index()
+        group_object.rename(index=str,
+                            columns={'DAYS_DECISION': 'previous_application_days_decision_about_prev_credit'},
+                            inplace=True)
+        features = features.merge(group_object, on=['SK_ID_CURR'], how='left')
+
+        group_object = prev_applications_last_3.groupby(by=['SK_ID_CURR'])['DAYS_DECISION'].mean().reset_index()
+        group_object.rename(index=str,
+                            columns={'DAYS_DECISION': 'previous_application_days_decision_about_last_3_credit_mean'},
```

Comment:
@pknut I really like this feature (if I understand it that is). Ok so this is the amount of time that some financial institution (or just home-credit?) had taken before they aproved the loan? If so it makes a lot of sense as a feature